eng/Npm.Workspace.nodeproj

```diff
@@ -19,15 +19,131 @@
     <_NpmAdditionalEnvironmentVariables>PUPPETEER_SKIP_DOWNLOAD=1</_NpmAdditionalEnvironmentVariables>
   </PropertyGroup>
 
-  <Target Name="Restore">
+  <PropertyGroup>
+    <NpmWorkspaceRoot>$(MSBuildThisFileDirectory)..</NpmWorkspaceRoot>
+    <NodeModulesMarkerFile>$(NpmWorkspaceRoot)\node_modules\.package-lock.json</NodeModulesMarkerFile>
+    <!-- Common exclude patterns for NPM build input/output files -->
+    <NpmCommonExcludePatterns>
+      **\pkg-version.ts;
+      $(NpmWorkspaceRoot)\src\**\node_modules\**\*;
+      $(NpmWorkspaceRoot)\src\**\bin\**\*;
+      $(NpmWorkspaceRoot)\src\**\obj\**\*
+    </NpmCommonExcludePatterns>
+  </PropertyGroup>
+
+  <!-- Input files for incremental restore check -->
+  <ItemGroup>
+    <NpmInputFiles Include="$(NpmWorkspaceRoot)\package.json" />
+    <NpmInputFiles Include="$(NpmWorkspaceRoot)\package-lock.json" Condition="Exists('$(NpmWorkspaceRoot)\package-lock.json')" />
+    <NpmInputFiles Include="$(NpmWorkspaceRoot)\src\**\package.json" Exclude="$(NpmWorkspaceRoot)\src\**\node_modules\**\package.json" />
+    <NpmInputFiles Include="$(NpmWorkspaceRoot)\.npmrc" Condition="Exists('$(NpmWorkspaceRoot)\.npmrc')" />
+  </ItemGroup>
+
+  <!-- Output files for incremental restore check -->
+  <ItemGroup>
+    <NpmOutputFiles Include="$(NodeModulesMarkerFile)" />
+  </ItemGroup>
+
+  <!-- Output files for incremental build check -->
+  <ItemGroup>
```

Comment:
How often might we have to update this list?

Reply:
We haven’t changed anything in years I would say.

The list of outputs is not as important as the incrementalism is “all or nothing” for all the us projects.

MSBUILD requires at least one output or else it will not run the task.

The goal is that if you’re not touching the JavaScript. You don’t have to build any part of the JavaScript pipeline during your dev builds.